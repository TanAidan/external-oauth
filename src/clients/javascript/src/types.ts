declare namespace Types {
  export interface ConnectOptions {
    authId?: string
    setupId?: string
    params?: any
  }

  export interface ConnectSuccess {
    authId: string
  }

  export interface ConnectError extends Error {}

  export interface IntegrationOptions {
    authId?: string
    setupId?: string
  }

  export type RequestHeaders = Record<string, string | number | undefined>
  export type RequestMethod = 'GET' | 'DELETE' | 'HEAD' | 'OPTIONS' | 'POST' | 'PUT' | 'PATCH'

  export type RequestQueryString = Record<string, string | number>
  export interface RequestParameters {
    headers?: RequestHeaders
    query?: RequestQueryString
    body?: any
  }
}

export default Types
