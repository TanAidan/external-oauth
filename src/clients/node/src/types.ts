declare namespace Types {
  export interface IntegrationOptions {
    authId?: string
    setupId?: string
  }

  export type RequestHeaders = Record<string, string | number | undefined>
  export type RequestMethod = 'GET' | 'DELETE' | 'HEAD' | 'OPTIONS' | 'POST' | 'PUT' | 'PATCH'

  export type RequestQueryString = Record<string, string | number>
  export interface RequestParameters {
    headers?: RequestHeaders
    query?: RequestQueryString
    body?: any
  }
}

export default Types
