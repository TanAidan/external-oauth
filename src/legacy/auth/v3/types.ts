import { Request } from 'express'
import Integration from '../../functions/integration'
import Knex from 'knex'

export interface IAuthContext {
  clientId: string
  connectParams: any
  buid?: string
  setupId: string
  isCallback?: boolean
  store: Knex
  configuration: SetupRecord
}

interface SetupRecord {
  setup_id: string
  scopes: string[] | undefined
  buid: string
  configuration: any
}

interface IContextSessionData {
  session: {
    context: any
  }
}

export enum EDashboardApiErrorCodes {
  ENVIRONMENT_NOT_FOUND = 'ENVIRONMENT_NOT_FOUND',
  INTEGRATION_NOT_FOUND = 'INTEGRATION_NOT_FOUND'
}

export interface IDashboardApiResponse {
  errors: [
    {
      code: EDashboardApiErrorCodes
      message: string
    }
  ]
  data: {
    viewer: {
      canInvoke: boolean
      hasAccessTo: boolean
      clientId: string
      setupId: string
      buid: string
    }
  }
}

export type TConnectContextRequest = Request &
  IContextSessionData &
  IAuthContext & {
    params: {
      buid: string
    }
    query: {
      clientId: string
      setupId: string
    }
    integration?: Integration
    fullLogs?: boolean
  }

export type TCallbackContextRequest = Request & IContextSessionData & IAuthContext

export enum ELocalRedirectMethod {
  Inline = 'inline',
  Localhost = 'localhost'
}

type TNameValues = Record<string, any>

export type TIntegrationConfig = TNameValues & {
  authType: EAuthType
  config?: {
    scope?: string[]
    state?: any
  }
}

export interface IAuthConfig {
  integrationConfig: TIntegrationConfig
  setupDetails: TNameValues
}

export enum EAuthType {
  NoAuth = 'NO_AUTH',
  OAuth1 = 'OAUTH1',
  OAuth2 = 'OAUTH2'
}

interface IAuthConfigSessionData {
  session: {
    authConfig: IAuthConfig
  }
}

export type TConnectConfigRequest = Request & IAuthContext & IAuthConfig & IAuthConfigSessionData

export type TCallbackConfigRequest = Request & IAuthConfig & IAuthConfigSessionData

export interface IAuthId {
  authId: string
}

export type TAuthIdRequest = Request &
  IAuthContext &
  IAuthId & {
    session: {
      authId: string
      context: {
        setupId: string
      }
    }
  }

export interface IOAuth1Credentials {
  accessToken: string
  tokenSecret: string
  consumerKey: string
  consumerSecret: string
  expiresIn: number
}

export interface IOAuth2Credentials {
  accessToken: string
  refreshToken: string
  idToken: string
  idTokenJwt?: any
  expiresIn: number
}

export interface IWithAuthCredentials {
  credentials: IOAuth1Credentials | IOAuth2Credentials
  tokenResponse?: OAuthTokenResponse
}

export type TAuthenticateRequest = Request & IAuthContext & IAuthConfig & IWithAuthCredentials

export type AuthSuccessRequest = Request &
  IAuthContext &
  IAuthId &
  IWithAuthCredentials & { buid: string; integrationConfig?: TIntegrationConfig } & { store: Knex }

export type TRevokeRequest = Request & { buid: string }

export type TErrorHandlerRequest = Request & Partial<IAuthContext> & Partial<IAuthId>

export interface IFetchAuthDetailsParams {
  buid: string
  scopedUserDataTableName: string
  servicesTableName: string
  integration: Integration
  environmentIdentifier: string
  authId?: string
  setupId?: string
  configuration: any
  setupIdFromRequest?: boolean
  store: any
}

export interface BasicAuthDetails {
  username: string
  password: string
}

export interface ApiKeyAuthDetails {
  apiKey: string
}

export enum OAuth1SignatureMethod {
  HmacSha1 = 'HMAC-SHA1',
  RsaSha1 = 'RSA-SHA1',
  PlainText = 'PLAINTEXT'
}

interface OAuthAuthDetails {
  callbackParams: Record<string, string>
  tokenResponse?: OAuthTokenResponse
  updatedAt: number
}

export interface OAuth1AuthDetails extends OAuthAuthDetails {
  accessToken: string
  expiresIn: number
  tokenSecret: string
  consumerKey: string
  consumerSecret: string
  signatureMethod: 'HMAC-SHA1' | 'PLAINTEXT' | 'RSA-SHA1'
}

export interface OAuth2AuthDetails extends OAuthAuthDetails {
  accessToken: string
  callbackParams: Record<string, string>
  clientId: string
  clientSecret: string
  expiresIn: number
  idToken?: string
  idTokenJwt?: any
  refreshToken?: string
}

export interface OAuthTokenResponse {
  body: any
  headers: Record<string, string>
}

export type AuthDetails = BasicAuthDetails | ApiKeyAuthDetails | OAuth1AuthDetails | OAuth2AuthDetails | undefined
