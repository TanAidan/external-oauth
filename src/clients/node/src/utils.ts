import Types from './types'

/**
 * Helper to build a new URL from different params
 */
export function toURL(origin: string, baseURL: string, endpoint: string, queryString?: Types.RequestQueryString): URL {
  const removeLeadingSlash = (text: string) => {
    return text.replace(/^\//, '')
  }

  const removeTrailingSlash = (text: string) => {
    return text.replace(/\/$/, '')
  }

  const urlParts: string[] = []
  urlParts.push(removeTrailingSlash(origin))
  urlParts.push(removeLeadingSlash(removeTrailingSlash(baseURL)))
  urlParts.push(removeLeadingSlash(endpoint))

  const url = new URL(urlParts.join('/'))

  if (queryString) {
    Object.keys(queryString).forEach(key => url.searchParams.append(key, String(queryString[key])))
  }

  return url
}

/**
 * Helper to remove all undefined keys
 * @param obj {object}
 */
export function cleanHeaders(obj: Record<string, any>) {
  return Object.keys(obj).reduce((acc, key: string) => {
    if (obj[key] !== undefined) {
      acc[key] = obj[key]
    }
    return acc
  }, {} as Record<string, any>)
}
