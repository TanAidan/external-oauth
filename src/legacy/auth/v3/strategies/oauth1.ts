import http from 'http'
import passport from 'passport'
import { OAuth } from 'oauth'
import OAuth1Strategy from 'passport-oauth1'
import { NextFunction, Response } from 'express'
import { createNamespace } from 'cls-hooked'

import {
  TAuthenticateRequest,
  IOAuth1Credentials,
  IFetchAuthDetailsParams,
  TIntegrationConfig,
  OAuth1SignatureMethod
} from '../types'
import { AuthenticationFailed, InvalidAuthId } from '../errors'
// import { getSetupDetails, getAuth, TOAuth1Payload } from '../../../clients/integrations'
import { checkSetupIdConsistency } from './setup-id-consistency'

const responseData = createNamespace('BearerOAuth1ResponseData')

const originalCreateClient = (OAuth.prototype as any)._createClient
;(OAuth.prototype as any)._createClient = function(...args) {
  responseData.set('response', undefined)

  const request: http.ClientRequest = originalCreateClient.apply(this, args)

  request.once('response', response => {
    responseData.set('response', response)
  })

  return request
}

type VerifyCallback = (err?: Error | null, user?: object, info?: object) => void

type VerifyFunction = (
  accessToken: string,
  refreshToken: string,
  params: any,
  profile: any,
  verified: VerifyCallback
) => void

interface StrategyOptions {
  consumerKey: string
  consumerSecret: string
  requestTokenURL: string
  tokenParams?: any
  accessTokenURL: string
  userAuthorizationURL: string
  authorizationParams?: any
  signatureMethod?: string
  callbackURL: string
  customHeaders?: any
}

class Strategy extends OAuth1Strategy {
  private _authorizationParams: any
  private _tokenParams: any

  constructor(options: StrategyOptions, verify: VerifyFunction) {
    super(options, verify)
    this._authorizationParams = options.authorizationParams
    this._tokenParams = options.tokenParams
  }

  userAuthorizationParams() {
    return this._authorizationParams || {}
  }

  requestTokenParams() {
    return this._tokenParams || {}
  }

  parseErrorResponse(body: string, statusCode: number) {
    return new AuthenticationFailed({ statusCode, body })
  }
}

const strategyOptions = (req: TAuthenticateRequest) => {
  const callbackURL = process.env.AUTH_CALLBACK_URL || `${req.protocol}://${req.get('host')}/auth/callback`
  const { consumerKey, consumerSecret } = req.setupDetails.credentials
  const {
    requestTokenURL,
    tokenParams,
    accessTokenURL,
    userAuthorizationURL,
    authorizationParams,
    signatureMethod
  } = req.integrationConfig
  return {
    consumerKey,
    consumerSecret,
    requestTokenURL,
    tokenParams,
    accessTokenURL,
    userAuthorizationURL,
    authorizationParams,
    signatureMethod,
    callbackURL
  }
}

const parseExpires = (expiresIn?: string) => {
  if (!expiresIn) {
    return 0
  }

  const result = parseInt(expiresIn, 10)

  if (isNaN(result)) {
    return 0
  }

  return result
}

export const authenticate = (req: TAuthenticateRequest, res: Response, next: NextFunction) => {
  // This is invoked after we've been called back by the third party, and when Passport
  // thinks we've succesfully got a token
  const verify = (accessToken: string, tokenSecret: string, params: any, _profile: any, verified: VerifyCallback) => {
    if (!accessToken) {
      return verified(undefined, undefined, { message: 'No access token returned', response: params })
    }
    const { consumerKey, consumerSecret } = req.setupDetails

    const credentials: IOAuth1Credentials = {
      consumerKey,
      consumerSecret,
      accessToken,
      tokenSecret,
      expiresIn: parseExpires(params.expires_in)
    }

    // The OAuth library removes the token/secret from the body data.
    // Add them back in so we can store the response data as it was originally received
    const body = {
      ...params,
      oauth_token: accessToken,
      oauth_token_secret: tokenSecret
    }

    verified(undefined, { body, credentials })
  }

  // This is invoked at the end of the Passport process when either an error has occurred,
  // authentication failed, or we have successfully authenticated. This might have been invoked
  // as a consequence of the `verified` callback above, or by Passport internally
  const authenticateCallback = (err: any, data: CallbackData | undefined, info: any) => {
    if (err) {
      return next(err)
    }

    if (!data) {
      return next(AuthenticationFailed.fromOAuthRequest(req, info))
    }

    const { body, credentials } = data

    req.credentials = credentials
    req.tokenResponse = {
      body,
      headers: responseData.get('response').headers
    }

    next()
  }

  responseData.run(() => {
    passport.use(new Strategy(strategyOptions(req), verify) as any)
    passport.authenticate('oauth', authenticateCallback)(req, res, next)
  })
}

export const nullAuthDetails = {
  accessToken: null,
  tokenSecret: null,
  consumerKey: null,
  consumerSecret: null,
  signatureMethod: null
}

export const fetchAuthDetails = async (params: IFetchAuthDetailsParams, integrationConfig: TIntegrationConfig) => {
  const {
    buid,
    // servicesTableName,
    // scopedUserDataTableName,
    // environmentIdentifier,
    // integration,
    authId,
    setupIdFromRequest,
    setupId: setupIdParam
  } = params

  const { signatureMethod = OAuth1SignatureMethod.HmacSha1 } = integrationConfig

  const credentials = {
    accessToken: 'accessToken',
    callbackParamsJSON: '{}',
    connectParams: {},
    expiresIn: 1234567890,
    tokenSecret: 'tokenSecret',
    setupId: 'setupId',
    updatedAt: 1234567890,
    consumerKey: 'consumerKey',
    consumerSecret: 'consumerSecret'
  }
  // const credentials = await getAuth<TOAuth1Payload>({
  //   servicesTableName,
  //   buid: integration.buid,
  //   authId: authId!,
  //   clientId: environmentIdentifier
  // })

  if (!credentials || !credentials.accessToken) {
    throw new InvalidAuthId(buid, authId!)
  }

  const {
    accessToken,
    callbackParamsJSON,
    connectParams = {},
    expiresIn,
    tokenSecret,
    setupId,
    updatedAt
  } = credentials
  const { consumerKey, consumerSecret } = credentials

  checkSetupIdConsistency({ setupId, setupIdParam, setupIdFromRequest })

  const callbackParams = callbackParamsJSON ? JSON.parse(callbackParamsJSON) : undefined

  if (!consumerKey || !consumerSecret) {
    // tslint:disable-next-line:semicolon
    // ;({ consumerKey, consumerSecret } = await getSetupDetails({
    //   scopedUserDataTableName,
    //   buid: integration.buid,
    //   setupId: setupId!,
    //   clientId: environmentIdentifier
    // }))
  }

  return {
    accessToken,
    callbackParams,
    connectParams,
    expiresIn,
    tokenSecret,
    consumerKey,
    consumerSecret,
    signatureMethod,
    updatedAt
  }
}

interface CallbackData {
  body: any
  credentials: IOAuth1Credentials
}
