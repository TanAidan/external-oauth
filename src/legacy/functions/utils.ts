// Is the given object empty e.g. {}?
export function isEmpty(obj: Object) {
  return Object.getOwnPropertyNames(obj).length === 0
}
