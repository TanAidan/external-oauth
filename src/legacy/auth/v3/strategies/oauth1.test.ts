import nock from 'nock'
import { MiddlewareTestHarness, session } from '../../../../../tests/utils'
import { authenticate, fetchAuthDetails } from './oauth1'
import { TAuthenticateRequest, EAuthType, TIntegrationConfig, OAuth1SignatureMethod } from '../types'
import {
  // getSetupDetails,
  // getAuth,
  TOAuth1Payload
} from '../../clients/integrations'
import Integration from '../../../functions/integration'

jest.mock('../../../clients/integrations')

const requiredIntegrationConfig = {
  authType: EAuthType.OAuth1,
  requestTokenURL: 'https://example.com/requestToken',
  accessTokenURL: 'https://example.com/accessToken',
  userAuthorizationURL: 'https://example.com/auth',
  config: {}
}

const fullIntegrationConfig = {
  ...requiredIntegrationConfig,
  authorizationParams: {
    someAuthParam: 'test-ap'
  },
  tokenParams: {
    someTokenParam: 'test-tp'
  },
  signatureMethod: 'PLAINTEXT'
}

const setupDetails = {
  consumerKey: 'test-consumer-key',
  consumerSecret: 'test-consumer-secret'
}

describe('authenticate', () => {
  const setup = (integrationConfig: TIntegrationConfig = fullIntegrationConfig) =>
    new MiddlewareTestHarness<TAuthenticateRequest>({
      configureRequest: req => {
        req.integrationConfig = integrationConfig
        req.setupDetails = setupDetails
      },
      setupMiddlewares: [session()],
      testMiddleware: authenticate
    })

  const mockRequestTokenReq = (requestBody = 'someTokenParam=test-tp') =>
    nock('https://example.com')
      .post('/requestToken', requestBody, {
        reqheaders: {
          'User-Agent': 'Pizzly',
          Authorization: new RegExp(
            'OAuth oauth_callback="https%3A%2F%2Fexample\\.com%2Fv2%2Fauth%2Fcallback",' +
              'oauth_consumer_key="test-consumer-key",oauth_nonce="w+",' +
              'oauth_signature_method="PLAINTEXT",oauth_timestamp="d+",oauth_version="1\\.0",' +
              'oauth_signature="test-consumer-secret%26"'
          )
        }
      })
      .reply(200, 'oauth_token=test-request-token&oauth_token_secret=test-request-token-secret')

  describe('when connecting', () => {
    it('obtains a request token', async () => {
      const reqTokenScope = mockRequestTokenReq()

      await setup().get()

      reqTokenScope.done()
    })

    it('redirects to the authorization url with the correct parameters', async () => {
      mockRequestTokenReq()

      await setup()
        .get()
        .expect(302)
        .expect(
          'Location',
          'https://example.com/auth?oauth_token=test-request-token&' +
            'oauth_callback=https%3A%2F%2Fexample.com%2Fv2%2Fauth%2Fcallback&someAuthParam=test-ap'
        )
    })

    it('works when optional config is not present', async () => {
      mockRequestTokenReq('')

      await setup(requiredIntegrationConfig)
        .get()
        .expect(302)
        .expect(
          'Location',
          'https://example.com/auth?oauth_token=test-request-token&' +
            'oauth_callback=https%3A%2F%2Fexample.com%2Fv2%2Fauth%2Fcallback'
        )
    })
  })

  describe('when called back by the third party', () => {
    const callbackSetup = async (integrationConfig?: TIntegrationConfig, tokenRequestBody?: string) => {
      const test = setup(integrationConfig)

      // Simulate connect request - needed to setup correct session state
      mockRequestTokenReq(tokenRequestBody)
      await test.get().expect(302)

      return test
    }

    const mockAccessTokenReq = () =>
      nock('https://example.com').post('/accessToken', '', {
        reqheaders: {
          'User-Agent': 'Pizzly',
          Authorization: new RegExp(
            'OAuth oauth_consumer_key ="test-consumer-key",oauth_nonce="w+",' +
              'oauth_signature_method="PLAINTEXT",oauth_timestamp="d+",' +
              'oauth_token="test-request-token",oauth_verifier="test-verifier",oauth_version="1.0",' +
              'oauth_signature="test-consumer-secret%26test-request-token-secret"'
          )
        }
      })

    it('exchanges the code for tokens and stores them in the request', async () => {
      const test = await callbackSetup()
      mockAccessTokenReq().reply(
        200,
        'oauth_token=test-access-token&oauth_token_secret=test-access-token-secret&expires_in=1234'
      )

      await test
        .get()
        .query('oauth_token=test-request-token,oauth_verifier=test-verifier')
        .expect(200)

      expect(test.req.credentials).toEqual({
        accessToken: 'test-access-token',
        tokenSecret: 'test-access-token-secret',
        consumerKey: 'test-consumer-key',
        consumerSecret: 'test-consumer-secret',
        expiresIn: 1234
      })
    })

    it('stores the token response data on the request', async () => {
      const test = await callbackSetup()
      mockAccessTokenReq().reply(
        200,
        'oauth_token=test-access-token&oauth_token_secret=test-access-token-secret&expires_in=1234',
        { some: 'header-value' }
      )

      await test
        .get()
        .query('oauth_token=test-request-token,oauth_verifier=test-verifier')
        .expect(200)

      expect(test.req.tokenResponse).toEqual({
        body: {
          oauth_token: 'test-access-token',
          oauth_token_secret: 'test-access-token-secret',
          expires_in: '1234'
        },
        headers: {
          some: 'header-value'
        }
      })
    })

    it('sets a default value for the expiry time if it is not returned', async () => {
      const test = await callbackSetup()
      mockAccessTokenReq().reply(200, 'oauth_token=test-access-token&oauth_token_secret=test-access-token-secret')

      await test
        .get()
        .query('oauth_token=test-request-token,oauth_verifier=test-verifier')
        .expect(200)

      expect(test.req.credentials).toEqual({
        accessToken: 'test-access-token',
        tokenSecret: 'test-access-token-secret',
        consumerKey: 'test-consumer-key',
        consumerSecret: 'test-consumer-secret',
        expiresIn: 0
      })
    })

    it('uses the default value for the expiry time if the returned value is not a number', async () => {
      const test = await callbackSetup()
      mockAccessTokenReq().reply(
        200,
        'oauth_token=test-access-token&oauth_token_secret=test-access-token-secret&expires_in=invalid'
      )

      await test
        .get()
        .query('oauth_token=test-request-token,oauth_verifier=test-verifier')
        .expect(200)

      expect(test.req.credentials).toEqual({
        accessToken: 'test-access-token',
        tokenSecret: 'test-access-token-secret',
        consumerKey: 'test-consumer-key',
        consumerSecret: 'test-consumer-secret',
        expiresIn: 0
      })
    })

    it('works when optional config is not present', async () => {
      const test = await callbackSetup(requiredIntegrationConfig, '')
      mockAccessTokenReq().reply(200, 'oauth_token=test-access-token&oauth_token_secret=test-access-token-secret')

      await test
        .get()
        .query('oauth_token=test-request-token,oauth_verifier=test-verifier')
        .expect(200)

      expect(test.req.credentials).toEqual({
        accessToken: 'test-access-token',
        tokenSecret: 'test-access-token-secret',
        consumerKey: 'test-consumer-key',
        consumerSecret: 'test-consumer-secret',
        expiresIn: 0
      })
    })

    it('raises an AuthenticationFailed error if there is an error getting the access token', async () => {
      const test = await callbackSetup()
      mockAccessTokenReq().reply(403, 'oops')

      await test
        .get()
        .query('oauth_token=test-request-token,oauth_verifier=test-verifier')
        .expect(403)

      expect(test.err).toMatchSnapshot()
    })

    it('raises an AuthenticationFailed error if no access token was returned', async () => {
      const test = await callbackSetup()
      mockAccessTokenReq().reply(200, 'non_standard=error')

      await test
        .get()
        .query('oauth_token=test-request-token,oauth_verifier=test-verifier')
        .expect(403)

      expect(test.err).toMatchSnapshot()
    })
  })
})

describe('fetchAuthDetails', () => {
  const log = jest.fn()

  // const consumerKey = 'test-consumer-key'
  // const consumerSecret = 'test-consumer-secret'
  const accessToken = 'test-access-token'
  const tokenSecret = 'test-token-secret'
  const setupId = 'test-setup-id'

  const paramsBase: any = {
    buid: 'test-alias',
    scopedUserDataTableName: 'test-scoped-data-table',
    servicesTableName: 'test-services-table',
    environmentIdentifier: 'test-env-id',
    integration: new Integration('test-buid'),
    authId: 'test-authId',
    logger: { log }
  }

  const credentials = {
    accessToken,
    tokenSecret,
    setupId,
    connectParams: { connect: 'connect-param' },
    expiresIn: 1234,
    updatedAt: 1111
  }

  const callbackParams = { param: 'value' }

  const setup = ({
    withCallbackParams = false,
    withCredentials = true,
    withFullCredentials = false
  }: {
    withCallbackParams?: boolean
    withCredentials?: boolean
    withFullCredentials?: boolean
  } = {}) => {
    log.mockClear()
    // mocked(getSetupDetails)
    //   .mockResolvedValueOnce({ consumerKey, consumerSecret })
    //   .mockClear()

    const credentialsToUse = { ...credentials } as TOAuth1Payload

    if (withCallbackParams) {
      credentialsToUse.callbackParamsJSON = JSON.stringify(callbackParams)
    }

    if (withFullCredentials) {
      credentialsToUse.consumerKey = 'my-other-consumer-key'
      credentialsToUse.consumerSecret = 'my-other-consumer-secret'
    }

    // mocked(getAuth)
    //   .mockResolvedValueOnce(withCredentials ? credentialsToUse : undefined)
    //   .mockClear()
  }

  it('returns the details from the auth credentials and setup details', async () => {
    setup()

    const auth = await fetchAuthDetails({ ...paramsBase, setupId }, fullIntegrationConfig)

    expect(auth).toMatchSnapshot()
    // expect(getSetupDetails).toMatchSnapshot()
    // expect(getAuth).toMatchSnapshot()
  })

  it('defaults to the HMAC-SHA1 signature type', async () => {
    setup()

    const auth = await fetchAuthDetails({ ...paramsBase, setupId }, requiredIntegrationConfig)

    expect(auth.signatureMethod).toBe(OAuth1SignatureMethod.HmacSha1)
  })

  it('returns any callback params data', async () => {
    setup({ withCallbackParams: true })

    const auth = await fetchAuthDetails(paramsBase, requiredIntegrationConfig)

    expect(auth.callbackParams).toEqual(callbackParams)
  })

  describe("when the setupId used when connecting doesn't match the setupId sent by the user in the request", () => {
    it('throws an InconsistentSetupId error', async () => {
      setup()

      const params = { ...paramsBase, setupId: 'other-setup-id', setupIdFromRequest: true }

      await expect(fetchAuthDetails(params, requiredIntegrationConfig)).rejects.toMatchSnapshot()
    })
  })

  describe("when the setupId used when connecting doesn't match the default setupId", () => {
    it('logs a warning message about inconsistent setupId', async () => {
      setup()

      const params = { ...paramsBase, setupId: 'other-setup-id', setupIdFromRequest: false }

      await fetchAuthDetails(params, requiredIntegrationConfig)

      expect(log).toHaveBeenCalledWith(
        // tslint:disable-next-line:max-line-length
        'Warning: setupId inconsistency - setupId used for authorization setup differs from latest setupId stored in API dashboard'
      )
    })
  })

  describe('when no setupId parameter is present', () => {
    it("doesn't raise an error", async () => {
      setup()

      const auth = await fetchAuthDetails(paramsBase, requiredIntegrationConfig)

      expect(auth).toMatchObject({ accessToken })
    })
  })

  describe('when consumerKey consumerSecret can be found in the auth', () => {
    it('returns the consumerKey and consumerSecret from the authId', async () => {
      setup({ withFullCredentials: true })

      const auth = await fetchAuthDetails(paramsBase, requiredIntegrationConfig)

      expect(auth).toMatchObject({
        consumerKey: 'my-other-consumer-key',
        consumerSecret: 'my-other-consumer-secret'
      })
    })

    it('returns any callback params data', async () => {
      setup({ withFullCredentials: true, withCallbackParams: true })

      const auth = await fetchAuthDetails(paramsBase, requiredIntegrationConfig)

      expect(auth.callbackParams).toEqual(callbackParams)
    })
  })

  describe('when no auth details can be found for the authId', () => {
    it('raises an InvalidAuthId error', async () => {
      setup({ withCredentials: false })

      await expect(fetchAuthDetails(paramsBase, requiredIntegrationConfig)).rejects.toMatchInlineSnapshot(`
[InvalidAuthId: No auth details were found using authId 'test-authId' for API 'test-alias'

Please try again with a valid authId or connect with this authId]
`)
    })
  })
})
