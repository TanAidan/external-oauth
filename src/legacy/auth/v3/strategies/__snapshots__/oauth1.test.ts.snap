// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authenticate when called back by the third party raises an AuthenticationFailed error if no access token was returned 1`] = `
[AuthenticationFailed: Authentication failed:
{
  "message": "No access token returned",
  "response": {
    "non_standard": "error"
  }
}]
`;

exports[`authenticate when called back by the third party raises an AuthenticationFailed error if there is an error getting the access token 1`] = `
[AuthenticationFailed: Authentication failed:
{
  "statusCode": 403,
  "body": "oops"
}]
`;

exports[`fetchAuthDetails returns the details from the auth credentials and setup details 1`] = `
Object {
  "accessToken": "accessToken",
  "callbackParams": Object {},
  "connectParams": Object {},
  "consumerKey": "consumerKey",
  "consumerSecret": "consumerSecret",
  "expiresIn": 1234567890,
  "signatureMethod": "PLAINTEXT",
  "tokenSecret": "tokenSecret",
  "updatedAt": 1234567890,
}
`;

exports[`fetchAuthDetails when the setupId used when connecting doesn't match the setupId sent by the user in the request throws an InconsistentSetupId error 1`] = `[InconsistentSetupId: The provided setupId does not match the setupId used when connecting. Please reconnect using this setupId, or retry this call with the setupId corresponding to the authId]`;
